Add data model XML files for 1.2 spec

cecille opened the pull request:
#### Summary
After looking at some certification data, it is clear that we are not yet done with the older specification revisions. We're seeing a fair number of devices in production that are not compliant and we therefore need to back port some of the current tests to older revisions.

#### Testing
There are a set of unit tests that apply to the data model files - these have been updated in this PR. 
Testing against a 1.2 device is harder since we don't actually have any. We are forming a plan to get backwards compatibility testing running with the QA team.
I have manually audited the diff list below,but would appreciate some more eyes.

#### Readability checklist

Note that this PR is NOT short because this represents an entire set of new files from the spec. I have attempted to maintain the individual pieces in separate commits. The initial commit is a scrape from the 1.2 spec, subsequent commits add unit tests, build files and ALSO add some manual fixes where the 1.2 spec could not be scraped automatically by alchemy due to table formatting changes.
I have provided the diff below to indicate the differences from 1.2 to 1.3 (already in the repo)

#### Spec diff between 1.2 and 1.3 data model files
```
Clusters newly added in 1.3
['Boolean State Configuration', 'Content App Observer', 'Content Control', 'Device Energy Management', 'Electrical Energy Measurement', 'Electrical Power Measurement', 'Energy EVSE', 'Energy Preference', 'Laundry Dryer Controls', 'Messages', 'Microwave Oven Control', 'Device Energy Management Mode', 'Energy EVSE Mode', 'Microwave Oven Mode', 'Oven Mode', 'OTA Software Update Provider', 'OTA Software Update Requestor', 'Oven Cavity Operational State', 'Power Topology', 'Scenes Management', 'Valve Configuration and Control']


Clusters removed since 1.2
['Scenes', 'Leaf Wetness Measurement', 'Soil Moisture Measurement']


Cluster Identify
	Features removed: ['Query']
	Accepted Commands removed: ['IdentifyQuery']
	Generated Commands removed: ['IdentifyQueryResponse']


Cluster On/Off
	Revision change - old: 5 new: 6
	Features added: ['OffOnly']
	Features changed:
		Lighting
			LT: Lighting conformance O
			LT: Lighting conformance [!OFFONLY]
		DeadFrontBehavior (previously DeadFrontBehaviour)
			DF: DeadFrontBehaviour conformance O
			DF: DeadFrontBehavior conformance [!OFFONLY]
	Accepted Commands changed:
		On
			On id:0x01 1 conformance: M privilege: 3
			On id:0x01 1 conformance: !OFFONLY privilege: 3
		Toggle
			Toggle id:0x02 2 conformance: M privilege: 3
			Toggle id:0x02 2 conformance: !OFFONLY privilege: 3


Cluster Level Control
	Features changed:
		Frequency
			FQ: Frequency conformance O
			FQ: Frequency conformance P


Cluster Occupancy Sensing
	Revision change - old: 3 new: 4


Cluster Thermostat
	Features changed:
		ScheduleConfiguration (previously Schedule Configuration)
			SCH: Schedule Configuration conformance O
			SCH: ScheduleConfiguration conformance [Zigbee], D
	Attributes removed: ['AlarmMask']
	Attributes changed:
		ControlSequenceOfOperation
			ControlSequenceOfOperation: datatype: ThermostatControlSequence conformance: M, access = RW VM
			ControlSequenceOfOperation: datatype: ControlSequenceOfOperationEnum conformance: M, access = RW VM
		SystemMode
			SystemMode: datatype: ThermostatSystemMode conformance: M, access = RW VM
			SystemMode: datatype: SystemModeEnum conformance: M, access = RW VM
		ThermostatRunningMode
			ThermostatRunningMode: datatype: ThermostatSystemMode conformance: [AUTO], access = R V
			ThermostatRunningMode: datatype: ThermostatRunningModeEnum conformance: [AUTO], access = R V
		StartOfWeek
			StartOfWeek: datatype: StartOfWeek conformance: SCH, access = R V
			StartOfWeek: datatype: StartOfWeekEnum conformance: SCH, access = R V
		TemperatureSetpointHold
			TemperatureSetpointHold: datatype: TemperatureSetpointHold conformance: O, access = RW VM
			TemperatureSetpointHold: datatype: TemperatureSetpointHoldEnum conformance: O, access = RW VM
	Accepted Commands removed: ['GetRelayStatusLog']
	Generated Commands removed: ['GetRelayStatusLogResponse']


Cluster Pulse Width Modulation
	Features changed:
		Frequency
			FQ: Frequency conformance O
			FQ: Frequency conformance P


Cluster Basic Information
	Revision change - old: 2 new: 3
	Attributes added: ['SpecificationVersion', 'MaxPathsPerInvoke']


Cluster General Commissioning
	Accepted Commands added: ['ArmFailSafe', 'SetRegulatoryConfig', 'CommissioningComplete']
	Generated Commands added: ['ArmFailSafeResponse', 'SetRegulatoryConfigResponse', 'CommissioningCompleteResponse']


Cluster General Diagnostics
	Revision change - old: 1 new: 2
	Features added: ['DataModelTest']
	Attributes added: ['DoNotUse']
	Attributes changed:
		UpTime
			UpTime: datatype: uint64 conformance: O, access = R V
			UpTime: datatype: uint64 conformance: M, access = R V
	Accepted Commands added: ['TimeSnapshot', 'PayloadTestRequest']
	Generated Commands added: ['TimeSnapshotResponse', 'PayloadTestResponse']


Cluster Thread Network Diagnostics
	Revision change - old: 1 new: 2


Cluster Network Commissioning
	Revision change - old: 1 new: 2
	Attributes added: ['SupportedWiFiBands', 'SupportedThreadFeatures', 'ThreadVersion']
	Attributes changed:
		Networks
			Networks: datatype: UNKNOWN conformance: M, access = R A
			Networks: datatype: list conformance: M, access = R A


Cluster Bridged Device Basic Information
	Revision change - old: 2 new: 3
	Attributes added: ['SpecificationVersion', 'MaxPathsPerInvoke']


Cluster ICD Management
	Revision change - old: 1 new: 2
	Features added: ['UserActiveModeTrigger', 'LongIdleTimeSupport']
	Features changed:
		CheckInProtocolSupport
			CIP: CheckInProtocolSupport conformance O
			CIP: CheckInProtocolSupport conformance P, LITS, O
	Attributes added: ['UserActiveModeTriggerHint', 'OperatingMode']
	Accepted Commands changed:
		StayActiveRequest
			StayActiveRequest id:0x03 3 conformance: O privilege: 3
			StayActiveRequest id:0x03 3 conformance: LITS, O privilege: 3
	Generated Commands added: ['StayActiveResponse']


Cluster Laundry Washer Mode
	Revision change - old: 1 new: 2


Cluster Refrigerator And Temperature Controlled Cabinet Mode
	Revision change - old: 1 new: 2


Cluster RVC Run Mode
	Revision change - old: 1 new: 2
	Attributes changed:
		SupportedModes
			SupportedModes: datatype: list conformance: M, access = R V
			SupportedModes: datatype: list conformance: O, access = R V
		CurrentMode
			CurrentMode: datatype: uint8 conformance: M, access = R V
			CurrentMode: datatype: uint8 conformance: O, access = R V
		StartUpMode
			StartUpMode: datatype: uint8 conformance: O, access = RW VO
			StartUpMode: datatype: uint8 conformance: X, access = RW VO
		OnMode
			OnMode: datatype: uint8 conformance: DEPONOFF, access = RW VO
			OnMode: datatype: uint8 conformance: D, access = RW VO


Cluster RVC Clean Mode
	Revision change - old: 1 new: 2
	Attributes changed:
		SupportedModes
			SupportedModes: datatype: list conformance: M, access = R V
			SupportedModes: datatype: list conformance: O, access = R V
		CurrentMode
			CurrentMode: datatype: uint8 conformance: M, access = R V
			CurrentMode: datatype: uint8 conformance: O, access = R V
		StartUpMode
			StartUpMode: datatype: uint8 conformance: O, access = RW VO
			StartUpMode: datatype: uint8 conformance: X, access = RW VO
		OnMode
			OnMode: datatype: uint8 conformance: DEPONOFF, access = RW VO
			OnMode: datatype: uint8 conformance: D, access = RW VO


Cluster Refrigerator Alarm
	Attributes changed:
		Mask
			Mask: datatype: AlarmMap conformance: M, access = R V
			Mask: datatype: AlarmBitmap conformance: M, access = R V
		Latch
			Latch: datatype: AlarmMap conformance: RESET, access = R V
			Latch: datatype: AlarmBitmap conformance: RESET, access = R V
		State
			State: datatype: AlarmMap conformance: M, access = R V
			State: datatype: AlarmBitmap conformance: M, access = R V
		Supported
			Supported: datatype: AlarmMap conformance: M, access = R V
			Supported: datatype: AlarmBitmap conformance: M, access = R V


Cluster Dishwasher Mode
	Revision change - old: 1 new: 2


Cluster Dishwasher Alarm
	Attributes changed:
		Mask
			Mask: datatype: AlarmMap conformance: M, access = R V
			Mask: datatype: AlarmBitmap conformance: M, access = R V
		Latch
			Latch: datatype: AlarmMap conformance: RESET, access = R V
			Latch: datatype: AlarmBitmap conformance: RESET, access = R V
		State
			State: datatype: AlarmMap conformance: M, access = R V
			State: datatype: AlarmBitmap conformance: M, access = R V
		Supported
			Supported: datatype: AlarmMap conformance: M, access = R V
			Supported: datatype: AlarmBitmap conformance: M, access = R V


Cluster Operational State
	Revision change - old: 1 new: 2
	Events changed:
		OperationCompletion
			XmlEvent(name='OperationCompletion', conformance=optional(choice=None))
			XmlEvent(name='OperationCompletion', conformance=otherwise(choice=None))


Cluster RVC Operational State
	Revision change - old: 1 new: 2
	Accepted Commands added: ['GoHome']
	Accepted Commands changed:
		Pause
			Pause id:0x00 0 conformance: Resume, O privilege: 3
			Pause id:0x00 0 conformance: O privilege: 3
		Stop
			Stop id:0x01 1 conformance: Start, O privilege: 3
			Stop id:0x01 1 conformance: D privilege: 3
		Start
			Start id:0x02 2 conformance: O privilege: 3
			Start id:0x02 2 conformance: D privilege: 3
		Resume
			Resume id:0x03 3 conformance: Pause, O privilege: 3
			Resume id:0x03 3 conformance: O privilege: 3
	Generated Commands changed:
		OperationalCommandResponse
			OperationalCommandResponse id:0x04 4 conformance: Pause | Stop | Start | Resume privilege: None
			OperationalCommandResponse id:0x04 4 conformance: O privilege: None
	Events changed:
		OperationCompletion
			XmlEvent(name='OperationCompletion', conformance=optional(choice=None))
			XmlEvent(name='OperationCompletion', conformance=otherwise(choice=None))


Cluster Power Source
	Features changed:
		Replaceable
			REPLC: Replaceable conformance O
			REPLC: Replaceable conformance [BAT]
		Rechargeable
			RECHG: Rechargeable conformance O
			RECHG: Rechargeable conformance [BAT]
	Attributes changed:
		BatCapacity
			BatCapacity: datatype: uint32 conformance: [REPLC], access = R V
			BatCapacity: datatype: uint32 conformance: [REPLC | RECHG], access = R V


Cluster Door Lock
	Attributes changed:
		LockState
			LockState: datatype: enum8 conformance: M, access = R V
			LockState: datatype: LockStateEnum conformance: M, access = R V
		LockType
			LockType: datatype: enum8 conformance: M, access = R V
			LockType: datatype: LockTypeEnum conformance: M, access = R V
		DoorState
			DoorState: datatype: enum8 conformance: DPS, access = R V
			DoorState: datatype: DoorStateEnum conformance: DPS, access = R V
		CredentialRulesSupport
			CredentialRulesSupport: datatype: map8 conformance: USR, access = R V
			CredentialRulesSupport: datatype: CredentialRulesBitmap conformance: USR, access = R V
		LEDSettings
			LEDSettings: datatype: uint8 conformance: O, access = RW VM
			LEDSettings: datatype: LEDSettingEnum conformance: O, access = RW VM
		SoundVolume
			SoundVolume: datatype: uint8 conformance: O, access = RW VM
			SoundVolume: datatype: SoundVolumeEnum conformance: O, access = RW VM
		OperatingMode
			OperatingMode: datatype: enum8 conformance: M, access = RW VM
			OperatingMode: datatype: OperatingModeEnum conformance: M, access = RW VM
		SupportedOperatingModes
			SupportedOperatingModes: datatype: map16 conformance: M, access = R V
			SupportedOperatingModes: datatype: OperatingModesBitmap conformance: M, access = R V
		DefaultConfigurationRegister
			DefaultConfigurationRegister: datatype: map16 conformance: O, access = R V
			DefaultConfigurationRegister: datatype: ConfigurationRegisterBitmap conformance: O, access = R V
		LocalProgrammingFeatures
			LocalProgrammingFeatures: datatype: map8 conformance: O, access = RW VA
			LocalProgrammingFeatures: datatype: LocalProgrammingFeaturesBitmap conformance: O, access = RW VA
		AlarmMask
			AlarmMask: datatype: map16 conformance: O, access = RW VA
			AlarmMask: datatype: AlarmMaskBitmap conformance: O, access = RW VA
		KeypadOperationEventMask
			KeypadOperationEventMask: datatype: map16 conformance: [NOT & PIN], access = RW VA
			KeypadOperationEventMask: datatype: EventMaskBitmap conformance: [NOT & PIN], access = RW VA
		RemoteOperationEventMask
			RemoteOperationEventMask: datatype: map16 conformance: [NOT], access = RW VA
			RemoteOperationEventMask: datatype: EventMaskBitmap conformance: [NOT], access = RW VA
		ManualOperationEventMask
			ManualOperationEventMask: datatype: map16 conformance: [NOT], access = RW VA
			ManualOperationEventMask: datatype: EventMaskBitmap conformance: [NOT], access = RW VA
		RFIDOperationEventMask
			RFIDOperationEventMask: datatype: map16 conformance: [NOT & RID], access = RW VA
			RFIDOperationEventMask: datatype: EventMaskBitmap conformance: [NOT & RID], access = RW VA
		KeypadProgrammingEventMask
			KeypadProgrammingEventMask: datatype: map16 conformance: [NOT & PIN], access = RW VA
			KeypadProgrammingEventMask: datatype: EventMaskBitmap conformance: [NOT & PIN], access = RW VA
		RemoteProgrammingEventMask
			RemoteProgrammingEventMask: datatype: map16 conformance: [NOT], access = RW VA
			RemoteProgrammingEventMask: datatype: EventMaskBitmap conformance: [NOT], access = RW VA
		RFIDProgrammingEventMask
			RFIDProgrammingEventMask: datatype: map16 conformance: [NOT & RID], access = RW VA
			RFIDProgrammingEventMask: datatype: EventMaskBitmap conformance: [NOT & RID], access = RW VA


Cluster Wake On LAN
	Attributes changed:
		MACAddress
			MACAddress: datatype: hwadr conformance: O, access = R V
			MACAddress: datatype: string conformance: O, access = R V


Cluster Channel
	Revision change - old: 1 new: 2
	Features added: ['ElectronicGuide', 'RecordProgram']
	Accepted Commands added: ['GetProgramGuide', 'RecordProgram', 'CancelRecordProgram']
	Generated Commands added: ['ProgramGuideResponse']


Cluster Target Navigator
	Revision change - old: 1 new: 2
	Events added: ['TargetUpdated']


Cluster Media Playback
	Revision change - old: 1 new: 2
	Features added: ['TextTracks', 'AudioTracks', 'AudioAdvance']
	Attributes added: ['ActiveAudioTrack', 'AvailableAudioTracks', 'ActiveTextTrack', 'AvailableTextTracks']
	Accepted Commands added: ['ActivateAudioTrack', 'ActivateTextTrack', 'DeactivateTextTrack']
	Events added: ['StateChanged']


Cluster Content Launcher
	Revision change - old: 1 new: 2
	Features added: ['AdvancedSeek', 'TextTracks', 'AudioTracks']


Cluster Account Login
	Revision change - old: 1 new: 2
	Generated Commands changed:
		GetSetupPINResponse
			GetSetupPINResponse id:0x01 1 conformance: M privilege: None
			GetSetupPINResponse id:0x01 1 conformance: M privilege: 0
	Events added: ['LoggedOut']


Device Types newly added in 1.3
['Cook Surface', 'Cooktop', 'Device Energy Management', 'Energy EVSE', 'Electrical Sensor', 'Extractor Hood', 'Laundry Dryer', 'Microwave Oven', 'Oven', 'Rain Sensor', 'Water Freeze Detector', 'Water Leak Detector', 'Water Valve']


Device Types removed since 1.2
['Heating/Cooling Unit']


Device Type On/Off Light
	Revision change - old: 2 new: 3
	Server Clusters added: ['Scenes Management']
	Server Clusters removed: ['Scenes']


Device Type Dimmable Light
	Revision change - old: 2 new: 3
	Server Clusters added: ['Scenes Management']
	Server Clusters removed: ['Scenes']


Device Type Window Covering
	Revision change - old: 2 new: 3
	Server Clusters added: ['Scenes Management']
	Server Clusters removed: ['Scenes']


Device Type On/Off Light Switch
	Revision change - old: 2 new: 3
	Client Clusters added: ['Scenes Management']
	Client Clusters removed: ['Scenes']


Device Type Dimmer Switch
	Revision change - old: 2 new: 3
	Client Clusters added: ['Scenes Management']
	Client Clusters removed: ['Scenes']


Device Type Color Dimmer Switch
	Revision change - old: 2 new: 3
	Client Clusters added: ['Scenes Management']
	Client Clusters removed: ['Scenes']


Device Type Occupancy Sensor
	Client Clusters removed: ['Groups']


Device Type Door Lock
	Revision change - old: 2 new: 3
	Server Clusters removed: ['Alarms', 'Poll Control']
	Server Clusters changed:
		Groups
			Groups: [Zigbee]
			Groups: X
		Scenes Management (previously Scenes)
			Scenes: [Zigbee], P
			Scenes Management: X
	Client Clusters removed: ['Time']


Device Type Dimmable Plug-In Unit
	Revision change - old: 2 new: 4
	Server Clusters added: ['Scenes Management']
	Server Clusters removed: ['Scenes']
	Client Clusters added: ['Occupancy Sensing']


Device Type Color Temperature Light
	Revision change - old: 2 new: 4
	Server Clusters added: ['Scenes Management']
	Server Clusters removed: ['Scenes']
	Client Clusters added: ['Occupancy Sensing']


Device Type Door Lock Controller
	Revision change - old: 2 new: 3
	Server Clusters removed: ['Identify']
	Server Clusters changed:
		Time Synchronization (previously TimeSync)
			TimeSync: P, O
			Time Synchronization: O
	Client Clusters removed: ['Identify']
	Client Clusters changed:
		Groups
			Groups: Zigbee
			Groups: O
		Scenes Management (previously Scenes)
			Scenes: Zigbee, P
			Scenes Management: P, O


Device Type Extended Color Light
	Revision change - old: 2 new: 4
	Server Clusters added: ['Scenes Management']
	Server Clusters removed: ['Scenes']
	Client Clusters added: ['Occupancy Sensing']


Device Type Generic Switch
	Revision change - old: 2 new: 3


Device Type On/Off Plug-in Unit
	Revision change - old: 2 new: 3
	Server Clusters added: ['Scenes Management']
	Server Clusters removed: ['Scenes']
	Client Clusters added: ['Occupancy Sensing']


Device Type OTA Requestor
	Server Clusters added: ['OTA Software Update Requestor']
	Server Clusters removed: ['OTA Software Update Requestor']
	Client Clusters added: ['OTA Software Update Provider']
	Client Clusters removed: ['OTA Software Update Provider']


Device Type OTA Provider
	Server Clusters added: ['OTA Software Update Provider']
	Server Clusters removed: ['OTA Software Update Provider']
	Client Clusters added: ['OTA Software Update Requestor']
	Client Clusters removed: ['OTA Software Update Requestor']


Device Type Contact Sensor
	Revision change - old: 1 new: 2
	Server Clusters added: ['Boolean State Configuration']


Device Type Root Node
	Server Clusters changed:
		ICD Management
			ICD Management: SIT | LIT
			ICD Management: SIT


Device Type Thermostat
	Revision change - old: 2 new: 3
	Server Clusters added: ['Scenes Management', 'Energy Preference']
	Server Clusters removed: ['Scenes']
	Server Clusters changed:
		Time Synchronization (previously TimeSync)
			TimeSync: P, O
			Time Synchronization: O
	Client Clusters changed:
		Fan Control
			Fan Control: P, O
			Fan Control: O
		Time Synchronization (previously TimeSync)
			TimeSync: P, O
			Time Synchronization: O


Device Type Window Covering Controller
	Revision change - old: 2 new: 3
	Client Clusters added: ['Scenes Management']
	Client Clusters removed: ['Scenes']


Device Type Casting Video Player
	Revision change - old: 1 new: 2
	Server Clusters added: ['Messages', 'Content Control']


Device Type Content App
	Revision change - old: 1 new: 2
	Client Clusters added: ['Content App Observer']


Device Type Basic Video Player
	Revision change - old: 1 new: 2
	Server Clusters added: ['Messages', 'Content Control']


Device Type Casting Video Client
	Revision change - old: 1 new: 2
	Server Clusters added: ['Content App Observer']
	Client Clusters added: ['Messages', 'Content Control']


Device Type Pump
	Revision change - old: 2 new: 3
	Server Clusters added: ['Scenes Management']
	Server Clusters removed: ['Scenes']


Device Type Fan
	Revision change - old: 1 new: 2


Device Type Pump Controller
	Revision change - old: 2 new: 3
	Client Clusters added: ['Scenes Management']
	Client Clusters removed: ['Scenes']


Device Type Video Remote Control
	Revision change - old: 1 new: 2
	Client Clusters added: ['Content Control']


Device Type Control Bridge
	Revision change - old: 2 new: 3
	Client Clusters added: ['Scenes Management']
	Client Clusters removed: ['Scenes']


Device Type On/Off Sensor
	Revision change - old: 2 new: 3
	Client Clusters added: ['Scenes Management']
	Client Clusters removed: ['Scenes']


Device Type Temperature Controlled Cabinet
	Revision change - old: 1 new: 2
	Server Clusters added: ['Oven Cavity Operational State', 'Oven Mode']
	Server Clusters changed:
		Refrigerator and Temperature Controlled Cabinet Mode (previously Refrigerator And Temperature Controlled Cabinet Mode)
			Refrigerator And Temperature Controlled Cabinet Mode: O
			Refrigerator and Temperature Controlled Cabinet Mode: [Cooler]


Device Type Room Air Conditioner
	Revision change - old: 1 new: 2
	Server Clusters added: ['Scenes Management', 'Thermostat User Interface Configuration']
	Server Clusters removed: ['Scenes']


Device Type Robotic Vacuum Cleaner
	Revision change - old: 1 new: 2
```

cecille committed:
add files

cecille committed:
Add 1.2 to python whl

cecille committed:
Add prebuilt for 1.2

cecille committed:
Add new 1.2 files to python build

cecille committed:
huge hack - add base device clusters

Pre-1.3, the tables for these were divided into zigbee and matter.
Alchemy doesn't know about the older tables.

Given that this spec is stable now, I'm doing this manually from
the later revisions, since the revisions for this device type
has not changed.

We could write this into alchemy, but the effort / benefit ratio
there seems low.

cecille committed:
Door lock controller: Add conformance markers

These are marked as EZ-target and EZ-initiator, but those
conditions aren't exposed in a way that alchemy understands.
These are both zigbee specific conditions, and they are marked
in the spec as [EZ-Initiator] and [EZ-Target]. This means they're
optional only for zigbee and disallowed for matter.

In later revisions, identify is removed from this device type.

cecille committed:
Add 1.2 to device type unit tests

cecille committed:
Add 1.2 to selection and cluster tests, fix selection

tcarmelveilleux commented on src/python_testing/TestSpecParsingSelection.py:
```diff
@@ -69,8 +74,14 @@ def test_dm_from_spec_version(self):
-        xml_clusters, _ = build_xml_clusters(PrebuiltDataModelDirectory.k1_4_1)
+        if spec_version is None:
+            dm = PrebuiltDataModelDirectory.k1_2
+        elif spec_version == 0xFFFFFFFF:
+            # use 1.4.1, but change the spec version to something bad.
+            dm = PrebuiltDataModelDirectory.k1_4_1
```
Should this move to 1.4.2 now?

cecille replied:
1.4.2 isn't certified yet.

cecille replied:
I've been leaving it at the most up to date certified spec, which is 1.4.1 now. So I'd prefer to leave this as-is for now.

tcarmelveilleux commented on data_model/1.2/device_types/OnOffLight.xml:
```diff
@@ -0,0 +1,148 @@
+    <cluster id="0x0008" name="Level Control" side="server">
+      <optionalConform/>
+      <features>
+        <feature code="OO" name="OO">
+          <mandatoryConform/>
+        </feature>
+        <feature code="LT" name="LT">
+          <mandatoryConform/>
+        </feature>
```
I expected the "code" field to have the ID of the feature map bit, otherwise there is a double indirection needed to lookup the feature.

cecille replied:
If you want to suggest an update to the schema and discuss with Hasty, that's fine, but this is what we have now. It's a fairly direct representation of what's in the spec, which is intended to be human readable. So we'll need to do the lookup somewhere, it's just a question of whether we do it in alchemy or in the tests. The nice part of this schema is that it's readable by a human. Not that humans tend to do that too much. Current comment aside.

tcarmelveilleux commented on data_model/1.2/device_types/DimmableLight.xml:
```diff
@@ -0,0 +1,148 @@
+    <revision revision="1" summary="Initial Zigbee 3.0 release"/>
+    <revision revision="2" summary="New data model format and notation"/>
+  </revisionHistory>
+  <classification superset="On/Off Light" class="simple" scope="endpoint"/>
```
Having supersets by name rather than by ID is something that I think can lead to inconsistencies and definitely requires double-lookup. Does the scraper actually check every superset name is a valid device type name with an associated ID?

cecille commented on data_model/1.2/clusters/BallastConfiguration.xml:
```diff
@@ -0,0 +1,205 @@
+      <access read="true" write="true" readPrivilege="view" writePrivilege="manage"/>
+      <quality nullable="true"/>
+      <optionalConform/>
+      <constraint>
+        <between>
+          <from value="100"/>
+          <to value="MS"/>
+        </between>
```
Agree, but this comes from the spec, so I can't fix it here. Ballast configuration is removed in later spec revisions.

cecille commented on data_model/1.2/clusters/IlluminanceMeasurement.xml:
```diff
@@ -0,0 +1,137 @@
+      <access read="true" readPrivilege="view"/>
+      <quality nullable="true" reportable="true"/>
+      <mandatoryConform/>
+      <constraint>
+        <allowed value="0"/>
+      </constraint>
+      <constraint>
+        <between>
+          <from>
+            <attribute name="MinMeasuredValue"/>
+          </from>
+          <to>
+            <attribute name="MaxMeasuredValue"/>
+          </to>
+        </between>
+      </constraint>
```
It sure is confusing. This is what the spec says though, so nothing to be done here.

cecille commented on data_model/1.2/clusters/DoorLock.xml:
```diff
@@ -0,0 +1,2655 @@
+        </constraint>
+      </field>
+    </command>
+    <command id="0x08" name="Clear All PIN Codes" direction="commandToServer" response="Y">
+      <access invokePrivilege="admin" timed="true"/>
+      <mandatoryConform>
+        <andTerm>
+          <notTerm>
+            <feature name="USR"/>
+          </notTerm>
+          <feature name="PIN"/>
+        </andTerm>
+      </mandatoryConform>
+    </command>
```
yeah, door lock was rough in 1.2. This correctly reflects spec.

cecille commented on data_model/1.2/clusters/NetworkCommissioningCluster.xml:
```diff
@@ -0,0 +1,558 @@
+        <min value="1"/>
+      </constraint>
+    </attribute>
+    <attribute id="0x0001" name="Networks" default="empty">
+      <access read="true" readPrivilege="admin"/>
+      <mandatoryConform/>
+      <constraint>
+        <max>
+          <attribute name="MaxNetworks"/>
+        </max>
+      </constraint>
```
sure, I'll let the alchemy dev know.

cecille committed:
address review comments

cecille commented on src/python_testing/matter_testing_infrastructure/chip/testing/basic_composition.py:
```diff
@@ -237,8 +237,8 @@ def _get_dm(self) -> PrebuiltDataModelDirectory:  # type: ignore[return]
         except KeyError:
-            asserts.fail(
-                "Specification Version not found on device - ensure device bas a basic information cluster on EP0 supporting Specification Version")
+            # For now, assume we're looking at a 1.2 device (this is as close as we can get before the 1.1 and 1.0 DM files are populated)
```
added.

cecille committed:
Merge remote-tracking branch 'upstream/master' into dm_1.2

cecille committed:
Add missing OTA cluster files to 1.2

These are included weirdly in the 1.2 spec. Copy from 1.3. This is
a one-off, so I'm not fixing the scraper to deal with a spec
revision that's nearly sunset.

cecille committed:
manually add OTA clusters to 1.2 cluster id json

cecille committed:
Add new files to python build gn

cecille committed:
Update cluster IDs for OTA device types

Even though these clusters were missing anyway, that wasn't the
only problem - the device types used the IDs of the device
types rather than the clusters. This is all very fragile.
We need more spec checks.

cecille committed:
Update doc files for 1.2

restyled-commits committed:
Restyled by isort

cecille committed:
Merge remote-tracking branch 'upstream/master' into dm_1.2
